// @typegoose/typegoose/lib/internal/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StringLengthExpectedError = exports.FunctionCalledMoreThanSupportedError = exports.NotValidModelError = exports.RefOptionIsUndefinedError = exports.OptionRefDoesNotSupportArraysError = exports.SelfContainingClassError = exports.CannotBeSymbolError = exports.InvalidWhatIsItError = exports.AssertionFallbackError = exports.NoValidClassError = exports.NotAllVPOPElementsError = exports.NotStringTypeError = exports.NotNumberTypeError = exports.InvalidTypeError = void 0;
const utils_1 = require("./utils");
// Note: dont forget to use "toStringNoFail" on values that are "unknown" or "any"
class InvalidTypeError extends Error {
    constructor(targetName, key, Type) {
        super(`"${targetName}.${key}"'s Type is invalid! Type is: "${(0, utils_1.toStringNoFail)(Type)}" [E009]`);
    }
}
exports.InvalidTypeError = InvalidTypeError;
class NotNumberTypeError extends Error {
    constructor(targetName, key, enumKey, enumValue) {
        super(`Typeof "${targetName}.${key}" is "Number", value is undefined/null or does not have a reverse mapping! [E011]\n` +
            `  Encountered with property: "${enumKey}.${typeof enumValue}"`);
    }
}
exports.NotNumberTypeError = NotNumberTypeError;
class NotStringTypeError extends Error {
    constructor(targetName, key, enumKey, enumValue) {
        super(`Typeof "${targetName}.${key}" is "String", used enum is not only Strings! [E010]\n` +
            `  Encountered with property in Enum: "${enumKey}.${typeof enumValue}"`);
    }
}
exports.NotStringTypeError = NotStringTypeError;
/** Not All Virtual Populate Elements Error */
class NotAllVPOPElementsError extends Error {
    constructor(name, key) {
        super(`"${name}.${key}" has not all needed Virtual Populate Options! Needed are: ${utils_1.allVirtualoptions.join(', ')} [E006]`);
    }
}
exports.NotAllVPOPElementsError = NotAllVPOPElementsError;
class NoValidClassError extends TypeError {
    constructor(cl) {
        super(`"${(0, utils_1.toStringNoFail)(cl)}" is not a function(/constructor)!`);
    }
}
exports.NoValidClassError = NoValidClassError;
class AssertionFallbackError extends Error {
    constructor() {
        super('Assert failed - no custom error [E019]');
    }
}
exports.AssertionFallbackError = AssertionFallbackError;
/** Error for when an unknown WhatIsIt is passed to an switch, gets thrown in the default case */
class InvalidWhatIsItError extends Error {
    constructor(whatisit, name, key, where) {
        super(`"${(0, utils_1.toStringNoFail)(whatisit)}"(${where}) is invalid for "${name}.${key}" [E013]`);
    }
}
exports.InvalidWhatIsItError = InvalidWhatIsItError;
class CannotBeSymbolError extends Error {
    constructor(name, key) {
        super(`A property key in Typegoose cannot be an symbol! ("${name}.${(0, utils_1.toStringNoFail)(key)}") [E024]`);
    }
}
exports.CannotBeSymbolError = CannotBeSymbolError;
class SelfContainingClassError extends TypeError {
    constructor(name, key) {
        super('It seems like the type used is the same as the target class, which is not supported\n' +
            `Please look at https://github.com/typegoose/typegoose/issues/42 for more information ("${name}.${key}") [E004]`);
    }
}
exports.SelfContainingClassError = SelfContainingClassError;
class OptionRefDoesNotSupportArraysError extends TypeError {
    constructor(dim, name, key) {
        super(`Prop-Option "ref" does not support Arrays! (got "${dim}" dimensions, for property "${name}.${key}") [E021]`);
    }
}
exports.OptionRefDoesNotSupportArraysError = OptionRefDoesNotSupportArraysError;
class RefOptionIsUndefinedError extends Error {
    constructor(name, key) {
        super(`Prop-Option "ref"'s value is "null" or "undefined" for "${name}.${key}" [E005]`);
    }
}
exports.RefOptionIsUndefinedError = RefOptionIsUndefinedError;
class NotValidModelError extends TypeError {
    constructor(model, where) {
        super(`Expected "${where}" to be a valid mongoose.Model! (got: "${(0, utils_1.toStringNoFail)(model)}") [E025]`);
    }
}
exports.NotValidModelError = NotValidModelError;
class FunctionCalledMoreThanSupportedError extends Error {
    constructor(functionName, supported, extra) {
        super(`Function "${functionName}" only supports to be called "${supported}" times with the same parameters [E003]\n${extra}`);
    }
}
exports.FunctionCalledMoreThanSupportedError = FunctionCalledMoreThanSupportedError;
class StringLengthExpectedError extends TypeError {
    constructor(length, got, where, valueName) {
        // create the "got:" message, when string say it was a string, but not the length
        // if not string, then say it is not a string plus the value
        const gotMessage = typeof got === 'string' ? `(String: "${got.length}")` : `(not-String: "${(0, utils_1.toStringNoFail)(got)}")`;
        super(`Expected "${valueName}" to have at least length of "${length}" (got: ${gotMessage}, where: "${where}") [E026]`);
    }
}
exports.StringLengthExpectedError = StringLengthExpectedError;
